Backend api testing

54skyxenon opened the pull request:
Made a separate branch for unit testing the backend scaffold.

54skyxenon commented on api/src/test/java/com/codeforcommunity/rest/HttpConstantsTest.java:
```diff
@@ -0,0 +1,19 @@
+    String noteIdParam = "note_id";
+    String okMessage = "OK";
+    String noteRoute = "/api/note";
+} */
```
Just for consistency's sake, so that the `test` directory completely mirrors `main`.

connernilsen replied:
While you want to mirror `main`, you usually want your `test` directory to only contain the classes that either test or help test your classes in the main directory. The mirroring part becomes really useful when you want to test a package default/protected method, since the test directory allows you to have classes be in the same package without actually being in the `main` directory. Does that make sense?

54skyxenon commented:
Good news: I finish adding tests to RestFunctionsTest.java and they all work. @LukeBoyer  and I figured out how to unit test the routing through Mockito, so that we don't need a live API running. Also, I updated the README for testing and fixed your suggestions @connernilsen

connernilsen commented on api/src/test/java/com/codeforcommunity/rest/ApiRouterTest.java:
```diff
@@ -2,122 +2,101 @@
-  Router router;
+  INotesProcessor myINotesProcessor = Mockito.mock(INotesProcessor.class);
+  IAuthProcessor myIAuthProcessor = Mockito.mock(IAuthProcessor.class);
+  JWTAuthorizer myJWTAuthorizer = Mockito.mock(JWTAuthorizer.class);
```
This isn't something you have to change,  but if you want to you can statically import the `mockito.Mock` method (like `import static org.mockito.Mock`, which might make it slightly easier to mock classes. All you have to do at that point is `mock(<class>)`.

LukeBoyer commented on api/src/test/java/com/codeforcommunity/rest/ApiRouterTest.java:
```diff
@@ -0,0 +1,102 @@
+   * context.assertEquals(resp.statusCode(), 200);
+   * 
+   * context.assertEquals(b.containsKey("status"), true);
+   * context.assertEquals(b.containsKey("notes"), true);
```
Great start. Let's either delete commented out code or turn them into a functional unit test. Integration tests usually don't belong in test directory with unit tests. In practice integration tests are usually run from a separate service against a live instance of our application that's being exposed to actual user traffic, not usually against something running on someones personal machine used for development. Also not big on merging "todos" to master so let's either finish those or remove the files from the pr.

jackblanc commented:
Has this all been ported over to LLB now? Can we close this PR?

connernilsen commented:
@54skyxenon can this be closed or are you still planning on working on it?

54skyxenon commented:
@jackblanc  @connernilsen  I'm planning on continuing with it! Just took a long detour to support LLB

jackblanc commented:
@54skyxenon Reach out to me when you have time to start on this again! I'll be doing a lot of work in this repo these next few weeks

jackblanc commented:
@54skyxenon I'm going to close this PR, please reach out if you have time to work on this more

54skyxenon commented:
Sure thing, my focus right now is improving testing on the LLB backend.
